packages/opencensus-instrumentation-http/src/http.ts

```diff
@@ -0,0 +1,303 @@
+
+            rootSpan.addAttribute('http.host', request.url.host);
+            rootSpan.addAttribute('http.method', request.method);
+            rootSpan.addAttribute('http.path', request.url.pathname);
+            rootSpan.addAttribute('http.route', request.url.path);
+            rootSpan.addAttribute(
+                'http.user_agent', 'HTTPClient/' + request.httpVersion);
+            rootSpan.addAttribute('http.status_code', request.statusCode);
+
+            rootSpan.status = plugin.traceStatus(request.statusCode);
+
+            // Message Event ID is not defined
+            rootSpan.addMessageEvent(
+                'MessageEventTypeRecv', uuid.v4().split('-').join(''));
+
+            // debug('Ended root span: ', rootSpan.name, rootSpan.id,
+            // rootSpan.traceId);
+            rootSpan.end();
+            return returned;
+          };
+
+          return original.apply(this, arguments);
+        });
+      };
+    };
+  }
+
+  /**
+   * Creates spans for outgoing requests, sending spans' context for distributed
+   * tracing.
+   * @param plugin The HTTP plugin itself.
+   */
+  patchOutgoingRequest() {
+    const plugin: HttpPlugin = this;
+    // tslint:disable:no-any
+    return (original: any): any => {
+      return function outgoingRequest() {
+        // Makes sure the url is an url object
+        if (typeof (arguments[0]) === 'string') {
+          arguments[0] = url.parse(arguments[0]);
+        }
+
+        // Do not trace ourselves
+        if (arguments[0].headers &&
+            arguments[0].headers['x-opencensus-outgoing-request']) {
+          // tslint:disable:no-any
+          plugin.logger.debug(
+              'header with "x-opencensus-outgoing-request" - do not trace');
+          return original.apply(this, arguments);
+        }
+
+        const request = original.apply(this, arguments);
+
+        plugin.tracer.wrapEmitter(request);
+
+        plugin.logger.debug('%s plugin outgoingRequest', plugin.moduleName);
+        const traceOptions = {
+          name: `${request.method ? request.method : 'GET'}  ${
+              arguments[0].pathname}`,
+          type: 'CLIENT',
+        };
+
+
+        // Checks if this outgoing request is part of an operation by checking
+        // if there is a current root span, if so, we create a child span. In
+        // case there is no root span, this means that the outgoing request is
+        // the first operation, therefore we create a root span.
+        if (!plugin.tracer.currentRootSpan) {
+          plugin.logger.debug('outgoingRequest starting a root span');
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeRequestTrace(request, arguments, plugin));
+        } else {
+          plugin.logger.debug('outgoingRequest starting a child span');
+          const span: types.Span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (span: types.Span) =>
```

Comment:
The return type here is incorrect. This function should return `ClientRequest` but this is of type `(span: types.Span) => ClientRequest`.

Reply:
This is causing the `http` tests to fail, but not `https`. I believe the reason for this is because `http.get` and `https.get` are implemented differently; the way you are patching `https.get` in the `https` plugin, it is very likely that `req.end` is called _before_ this line, in `original.apply(this, arguments)` (line 170), because in Node 6 its implementation might simply call through the `http` implementation.

Reply:
Thanks @kjin.